<a href="https://colab.research.google.com/github/almsalahuddin/Data-Processing-and-Visualization/blob/main/Exam_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# EXAM 2 : CPSMA 4313 [Dr. Jacob's](https://github.com/nurfnick) Data processing visualization

### A L M SALAHUDDIN

Question(a) Store the data as a pandas dataframe. Examine each datatype and comment on the appropriateness of each.

##Import Libraries and Make Requests

First, Let's import the pandas library to read data from the web and then I will make request to collect all the data from given csv file

In [77]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/nurfnick/Data_Viz/main/Activity_Dataset_V1.csv')
df.head()

Unnamed: 0 activity_day    workout_type  distance  time  calories  \
0           0   2022-01-01       Freestyle      9.30    77       123   
1           1   2022-01-01       Freestyle      3.44    96        55   
2           2   2022-01-01  Indoor Cycling      6.34    85        33   
3           3   2022-01-01         Walking      7.91    42        82   
4           4   2022-01-01      Open Water      8.99    36       131   

   total_steps  avg_speed  avg_cadence  max_cadence  ... max_pace min_pace  \
0          NaN      18.88       168.54       138.30  ...      NaN      NaN   
1          NaN      29.65       125.92       292.81  ...      NaN      NaN   
2          NaN      17.85        81.93       323.69  ...      NaN      NaN   
3       1571.0      22.10        29.63       180.16  ...    28:58    07:58   
4          NaN      25.83        64.55       342.89  ...      NaN      NaN   

  avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0          112.5           122.0             103          19          28   
1          111.0           122.0             100          42          28   
2           95.0            90.0             100           1          32   
3           83.0            85.0              81           3          22   
4          138.0           166.0             110           7           0   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88  

[5 rows x 21 columns]

Next, I will find out all the datatypes.

In [78]:
df.dtypes

Unnamed: 0          int64
activity_day       object
workout_type       object
distance          float64
time                int64
calories            int64
total_steps       float64
avg_speed         float64
avg_cadence       float64
max_cadence       float64
avg_pace           object
max_pace           object
min_pace           object
avg_heart_rate    float64
max_heart_rate    float64
min_heart_rate      int64
vo2_max(%)          int64
aerobic(%)          int64
anaerobic(%)        int64
intensive(%)        int64
light(%)            int64
dtype: object

After analysing the dataset above, I would say datatype of activity_day should be DateTime datatype instead of object. Also, datatype of time, calories, and min_heart_rate should be in float datatype instead of int.

##(b) (10 points) Remove the column that repeats the indexes and is ‘unnamed’ as a column.

I will use drop function to remove unnamed column.

In [79]:
df = df.drop(columns=['Unnamed: 0'])
df.head()

activity_day    workout_type  distance  time  calories  total_steps  \
0   2022-01-01       Freestyle      9.30    77       123          NaN   
1   2022-01-01       Freestyle      3.44    96        55          NaN   
2   2022-01-01  Indoor Cycling      6.34    85        33          NaN   
3   2022-01-01         Walking      7.91    42        82       1571.0   
4   2022-01-01      Open Water      8.99    36       131          NaN   

   avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0      18.88       168.54       138.30      NaN      NaN      NaN   
1      29.65       125.92       292.81      NaN      NaN      NaN   
2      17.85        81.93       323.69      NaN      NaN      NaN   
3      22.10        29.63       180.16    07:58    28:58    07:58   
4      25.83        64.55       342.89      NaN      NaN      NaN   

   avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0           112.5           122.0             103          19          28   
1           111.0           122.0             100          42          28   
2            95.0            90.0             100           1          32   
3            83.0            85.0              81           3          22   
4           138.0           166.0             110           7           0   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88

##(c) (10 points) Clean the column names to remove the unit declaration, (%), using regular expressions. The column name should not have any trailing spaces after cleaning. You will only receive partial credit for simply renaming columns without using regular expressions.

Next, I will import re library and use lamda function to remove unit declaration(%).

In [80]:
import re
df = df.rename(columns = lambda x: re.sub(r'[(.%)]' ,'' ,x))
df.columns

Index(['activity_day', 'workout_type', 'distance', 'time', 'calories',
       'total_steps', 'avg_speed', 'avg_cadence', 'max_cadence', 'avg_pace',
       'max_pace', 'min_pace', 'avg_heart_rate', 'max_heart_rate',
       'min_heart_rate', 'vo2_max', 'aerobic', 'anaerobic', 'intensive',
       'light'],
      dtype='object')

Above we can see no more unit declaration(%) inside column.

##(d) (10 points) Convert ‘activity day’ column into a datetime format.

In order to convert convert ‘activity day’ column into a datetime format we have to know it's current datatype.

In [81]:
df.activity_day.dtype

dtype('O')

Now, I will conver the column into a datetime format 

In [82]:
pd.to_datetime(df.activity_day)

0     2022-01-01
1     2022-01-01
2     2022-01-01
3     2022-01-01
4     2022-01-01
         ...    
995   2022-02-14
996   2022-02-14
997   2022-02-14
998   2022-02-14
999   2022-02-14
Name: activity_day, Length: 1000, dtype: datetime64[ns]

Above, it's shows datatype converted to datetime format.

##(e) (10 points) Impute ‘total steps’ by replacing the ‘NaN’s with an appropriate number of steps. Convert to appropriate datatype.

First, lets find out how many ‘NaN’ value is there inside 'total steps’ column.

In [83]:
df.isnull().sum(axis = 0)

activity_day        0
workout_type        0
distance            0
time                0
calories            0
total_steps       446
avg_speed           0
avg_cadence         0
max_cadence         0
avg_pace          446
max_pace          446
min_pace          446
avg_heart_rate      0
max_heart_rate      0
min_heart_rate      0
vo2_max             0
aerobic             0
anaerobic           0
intensive           0
light               0
dtype: int64

Above we can see there is total of 446 'NaN' values inside total_steps. Next, I will find out it's datatype.

In [84]:
df.total_steps.dtype

dtype('float64')

Now, I will replace all the 'Nan' value using fillna function and all values will be replacing with the mean of total_steps.

In [85]:
df.total_steps.fillna(value=df.total_steps.mean(), inplace=True)
df.head()

activity_day    workout_type  distance  time  calories  total_steps  \
0   2022-01-01       Freestyle      9.30    77       123  4081.119134   
1   2022-01-01       Freestyle      3.44    96        55  4081.119134   
2   2022-01-01  Indoor Cycling      6.34    85        33  4081.119134   
3   2022-01-01         Walking      7.91    42        82  1571.000000   
4   2022-01-01      Open Water      8.99    36       131  4081.119134   

   avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0      18.88       168.54       138.30      NaN      NaN      NaN   
1      29.65       125.92       292.81      NaN      NaN      NaN   
2      17.85        81.93       323.69      NaN      NaN      NaN   
3      22.10        29.63       180.16    07:58    28:58    07:58   
4      25.83        64.55       342.89      NaN      NaN      NaN   

   avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max  aerobic  \
0           112.5           122.0             103       19       28   
1           111.0           122.0             100       42       28   
2            95.0            90.0             100        1       32   
3            83.0            85.0              81        3       22   
4           138.0           166.0             110        7        0   

   anaerobic  intensive  light  
0          2          7     50  
1          2         29     88  
2          0         22     43  
3          0         24     65  
4          5         21     88

Now, I will check agin to see if is there still any 'Nan' values left or I was able to remove all of them successfully.

In [86]:
df.isnull().sum(axis = 0)

activity_day        0
workout_type        0
distance            0
time                0
calories            0
total_steps         0
avg_speed           0
avg_cadence         0
max_cadence         0
avg_pace          446
max_pace          446
min_pace          446
avg_heart_rate      0
max_heart_rate      0
min_heart_rate      0
vo2_max             0
aerobic             0
anaerobic           0
intensive           0
light               0
dtype: int64

Above, it shows we don't have anymore 'Nan' values left for total_steps column.

##(f) (10 points) Convert non-empty ‘avg pace’ into a float that is still representative of the information contained in the column. Recall that there are 60 seconds in one minute so 3:30 is equivalent to 3.5 minutes.

In [87]:
df.avg_pace.dtype

dtype('O')

In [88]:
df.avg_pace = pd.DatetimeIndex(df.avg_pace)
df.avg_pace = df.avg_pace.apply(lambda avg_pace: avg_pace.hour + (avg_pace.minute / 60))
df.avg_pace.head(10)

0         NaN
1         NaN
2         NaN
3    7.966667
4         NaN
5         NaN
6         NaN
7    6.733333
8    6.800000
9         NaN
Name: avg_pace, dtype: float64

##(g) (10 points) Group data by ‘workout type’ and find the mean, median, count and standard deviation of ‘calories’.

I will find the mean, median, count and standard deviationof calories using agg function and put it groupby workout_type.

In [89]:
df.groupby('workout_type').calories.agg(['mean', 'median', 'count', 'std'])

mean  median  count         std
workout_type                                          
Cricket          307.548387   330.0     93  149.950609
Freestyle        278.552083   294.0     96  163.703768
Indoor Cycling   280.450000   306.0     80  156.633322
Open Water       296.747253   328.0     91  160.068499
Outdoor Cycling  299.129412   301.0     85  158.731263
Outdoor Running  301.469136   349.0     81  165.725080
Pool Swimming    283.414894   300.0     94  157.576703
Trail Run        267.966667   264.0     90  155.748533
Treadmill        278.142857   269.5     98  146.963352
Trekking         283.127660   280.5     94  151.034870
Walking          276.040816   270.5     98  157.287570

##(h) (10 points) Create an indicator column that identifies if the activity achieved 30% or more ‘aerobic’ activity.

I will import numphy liraries to call where function to decide whether the activity achieved 30% or more ‘aerobic’ activity and shows it under indicator column.

In [90]:
import numpy as np
df['indicator'] = np.where(df['aerobic'] >= 30, 'achived', 'not_achived')
df.head()

activity_day    workout_type  distance  time  calories  total_steps  \
0   2022-01-01       Freestyle      9.30    77       123  4081.119134   
1   2022-01-01       Freestyle      3.44    96        55  4081.119134   
2   2022-01-01  Indoor Cycling      6.34    85        33  4081.119134   
3   2022-01-01         Walking      7.91    42        82  1571.000000   
4   2022-01-01      Open Water      8.99    36       131  4081.119134   

   avg_speed  avg_cadence  max_cadence  avg_pace  ... min_pace avg_heart_rate  \
0      18.88       168.54       138.30       NaN  ...      NaN          112.5   
1      29.65       125.92       292.81       NaN  ...      NaN          111.0   
2      17.85        81.93       323.69       NaN  ...      NaN           95.0   
3      22.10        29.63       180.16  7.966667  ...    07:58           83.0   
4      25.83        64.55       342.89       NaN  ...      NaN          138.0   

   max_heart_rate  min_heart_rate  vo2_max  aerobic  anaerobic  intensive  \
0           122.0             103       19       28          2          7   
1           122.0             100       42       28          2         29   
2            90.0             100        1       32          0         22   
3            85.0              81        3       22          0         24   
4           166.0             110        7        0          5         21   

   light    indicator  
0     50  not_achived  
1     88  not_achived  
2     43      achived  
3     65  not_achived  
4     88  not_achived  

[5 rows x 21 columns]

##(i) (10 points) Which day of the week (Monday, Tuesday, etc.) and ‘workout type‘ has the maximum of the ‘max cadence’.

To solve this question, first we have to find out the maximum cadence and then find out what type of workout achived that max number and then what day it was when it was achived.

In [91]:
df1 = df.loc[df['max_cadence'] == df['max_cadence'].max()]
print('maximum_cadence = ' + str(df['max_cadence'].max()))
print('workout_type = ' + df1.workout_type.iloc[0])
print('day_of_week = ' + df1.activity_day.dt.day_name().iloc[0])

maximum_cadence = 349.78
workout_type = Pool Swimming


AttributeError: ignored